### ISSUE TITLE
Click on social login link in a new popup (which Cypress doesn't have access to)

### ISSUE THREAD

**Issue by michaelwschultz** [lirantal/cypress-social-logins#4]
Nice little plugin! This is going to be really helpful if I can get it working.

I've turned off headless so I can see what's happening. The google login window opens automatically but I can't seem to get it to fill in the creds. Any ideas?

![Screen Shot 2019-07-11 at 3 41 57 PM](https://user-images.githubusercontent.com/57668/61090231-a0ff2680-a3f2-11e9-9f0f-f5386ca3f7dd.png)

**Comment by lirantal** [comment]
@michaelwschultz thanks, let's try to figure this out.

can you post a more wholesome picture of what's going on or a video? since this is e2e it will help to get more context on where it fails.

also, can you manually trigger the `connect with google` window and then check if the google sign-in page has that email input type on the page with devtools?

**Comment by michaelwschultz** [comment]
@lirantal absolutely! I'll post a video soon. But yes, I've verified the input is on the google popup.

**Comment by michaelwschultz** [comment]
@lirantal video!

https://www.loom.com/share/0050194238a4488193a1009fd27288f2

**Comment by lirantal** [comment]
I can't spot anything bad there.
Tried to run it on a local site and the email part works well for me:

![image](https://user-images.githubusercontent.com/316371/61187890-20555a00-a680-11e9-995d-3b8106fa0c27.png)

Seems you already have the domain chosen while in my picture the email field is entirely empty, so perhaps that's a different form indeed and we need to tweak the selector.

I can suggest that in your own project where you have `cypress-social-logins` installed then open the `nodes_modules/` folder and the code for the plugin in `src/Plugin.js` directly of the `typeUsername` function to try a different selector, or perhaps see if the problem is elsewhere.

Happy to try out some other ideas with you too.

**Comment by michaelwschultz** [comment]
Yeah, I think it selects the field automatically when the window opens. Interesting that it doesn't do that for you. I'll try messing around with that function and see if I can't get it to work. Thanks for taking a look. I'll report back!

**Comment by raptatinha** [comment]
Hey @michaelwschultz, nice to meet you :)
I'm Renata and I had the same issue when I was passing wrongly the 'username' data to be typed in the code.
So, a few tips for you to double check:

1. If you are using a data file, make sure your data file is correct, here is an example of a loginData.js:
```
module.exports = {
    manager: { username: 'test@test.com', password: 'pwd' },
  };

```
in my login spec file login.spec.js i have something like this (**make sure you have no TYPO**):

```
const username = loginData.manager.username;

const socialLoginOptions = {
    username,
    password,
    loginUrl: Cypress.config().baseUrl,
    headless: true,
    logs: false,
    loginSelector: '[action="/Account/ExternalLogin?returnurl=/dashboard"] > [name="provider"]',
    postLoginSelector: '.p-dashboard',
  };

```

2. If you are not using a data file, but an env var, make sure the env is set correctly. And you are able to access it.

3. You could, as a last try, to use the data inside your code (at least to double check if it works and do a console.log do triple check)..
Something like this:

```
const username = 'test@test.com';
console.log('***** email: '+ username);

const socialLoginOptions = {
    username,
    password,
    loginUrl: Cypress.config().baseUrl,
    headless: true,
    logs: false,
    loginSelector: '[action="/Account/ExternalLogin?returnurl=/dashboard"] > [name="provider"]',
    postLoginSelector: '.p-dashboard',
  };

```

Happy trying!

**Comment by michaelwschultz** [comment]
Hey @raptatinha! Thank you so much for your help. I'll give this a shot later today. One question I have. How did you determine your loginSelector? It looks significantly different to what's in the template.

**Comment by lirantal** [comment]
the `loginSelector` comes from you, in the app you develop that is. That is basically the selector of the button/link/image on your own webpage that needs to be clicked in order to trigger the "Connect with Google" scenario.

**Comment by lirantal** [comment]
Let us know how it goes and feel free to brainstorm with us the issue. We'd love to help ❤️

**Comment by michaelwschultz** [comment]
@lirantal oh yeah, duh. Sorry I have that part working fine.

**Comment by zrana** [comment]
@lirantal 
I am also facing the same issue.
CypressError: cy.task('GoogleSocialLogin') failed with the following error:

> TimeoutError: waiting for selector "input[type="email"]" failed: timeout 30000ms exceeded

I had tried with different selector in src/plugin.js but getting the same error, any help?

**Comment by kashifch** [comment]
@lirantal @michaelwschultz 

This might help in debugging the issue, the problem is that the login page is opened in a separate chromium window and Cypress is able to click on the login button there, after that the email field appears in a new pop up and Cypress is unable to detect this pop up, when I copied the url of resulting popup and pasted it in original chromium page Cypress immediately picked email field

**Comment by id-dan** [comment]
Hi @michaelwschultz ! 
FYI: It is not good solution to use external pop-up for login via social networks because pop-up can be blocked by browser's settings.

For ex:
![image](https://user-images.githubusercontent.com/52101871/61808961-159b8180-ae45-11e9-94d5-b1ddc90c87de.png)

**Comment by lirantal** [comment]
@kashifch @zrana: to see if I understood correctly - you provide the plugin with the selector of the social login button to click on. This selector needs to be available on the DOM of the page you are testing with Cypress. If the Login button itself is in another popup then indeed this plugin can't access it.

However I don't think that's the case.
I think I was able to reproduce the problem now so looking at it.

P.S. @id-dan has a good point in general :)

**Comment by lirantal** [comment]
@kashifch 
So I think we nailed it indeed and it eluded me from your video - the fact that the Login button opens the login phase (user/pass etc) in a new popup is the culprit as the plugin has no idea where to look for it.

A couple of action items we can work on:
1. Are you able to change the experience to make the login work on the same tab by navigating to the social login? if so that's the easiest and fastest route 
2. If not, can you provide me the address for this login page so I can try and work out a feature support in this plugin that could allow it to work with popups? you're welcome to DM me on twitter to share it privately.

**Comment by zrana** [comment]
@lirantal @michaelwschultz 

I have forked the [repo](https://github.com/zrana/cypress-social-logins) and merged potential fix into master, please review. I have added a check which will switch to pop up if there is one and perform login operation there. otherwise it will stay on same page and add credentials.

**Comment by michaelwschultz** [comment]
@id-dan That's a great point! We should fix this in our auth config.

**Comment by michaelwschultz** [comment]
@zrana amazing! I'll take a look! @lirantal @id-dan @kashifch thanks for all your help! Excited to get this up and running. It's going to make all the rest of our end-to-end testing a breeze. 👍

**Comment by michaelwschultz** [comment]
@zrana really elegant solution. You should put up a PR on @lirantal's repo.

**Comment by lirantal** [comment]
@zrana definitely PR it to this repo, let's make sure we get it in.

I have some notes though on how you've got things there. Did you test the plugin without providing the target parameter? because I fear you'd hit a few undefined issues with how that code is currently written in your fork. Open up the PR and let's make it happen ;-)

**Comment by michaelwschultz** [comment]
@zrana any progress here? Also happy to open a PR with your code if you don't have time 😄

**Comment by lirantal** [comment]
@michaelwschultz will be great if you could send a PR here if @zrana won't be responsive.

**Comment by kashifch** [comment]
@lirantal 

@zrana is on leave this week, he will be back by Monday and then respond

**Comment by lirantal** [comment]
Sounds good, thanks for chiming in @kashifch 👍

**Comment by silverhand20** [comment]
Hi, I have a problem with Google Authentication.
That is, when I go to the page through the button (Login via Google). I am redirected to login from a Google account, but there are no account selection fields. Thanks for the help) 
https://www.dropbox.com/s/img9fvo79dhhe8k/t4VVl8Zkly.mp4?dl=0

**Comment by id-dan** [comment]
@silverhand20 Your case will not work because you tried to use multi-domains in the same test, so cypress can't do it - https://docs.cypress.io/guides/guides/web-security.html#One-Superdomain-per-Test

You should use this plugin - cypress-social-logins and it only will work in automation case (not manually as you tried) because puppeteer runs by **cy.task()** in separate headless window (not the same window as u tried)

So, pls carefully read documentation(examples)
https://github.com/lirantal/cypress-social-logins/blob/master/README.md

**Comment by silverhand20** [comment]
Thanks for the reply, but I still don't understand why it can't register my task?
Photo: https://i.imgur.com/jcgTFYv.png

**Comment by lirantal** [comment]
@michaelwschultz @kashifch @zrana any of you would like to pursue work on this with a PR? seems like you have a working solution in your own branch that only needs to be merged here?

**Comment by webcrew62** [comment]
@lirantal @id-dan @silverhand20
Can you please further clarify, perhaps with an example of multi-domain testing with cypress? For example, once authenticated to google, how can another domain be used in a cypress test?

**Comment by lirantal** [comment]
Cypress by design wouldn't work beyond the DOM so you can't control popups, iframes or other domains that are not the one you have Cypress running in. This plugin does it by launching puppeteer behind the scenes to do the login on the google domain.

**Comment by webcrew62** [comment]
@lirantal once the login on google domain is complete, the PostLoginSelector fails, as all elements on that page are signed. Is there a work-around for this?

**Comment by lirantal** [comment]
what does this mean: "as all elements on that page are signed" ?
once the login on google domain is complete, you are forwarded back to your own domain which should detect you as signed in, and then you can take the cookies and do with them whatever is needed in the Cypress testing part (such as logging you in).

**Comment by webcrew62** [comment]
Please disregard the previous comment about signed elements.

What (url) in the cypress.json is to indicate the domain which detects being signed in to google?

**Comment by lirantal** [comment]
So it's not exactly the domain that detects you are signed in to google.
Here is a reference to a cypress.json definition:

```
{
  "baseUrl": "https://example.com",
  "env": {
    "cookieName": "website.cookie.id",
    "googleSocialLoginUsername": "some@gmail.com",
    "googleSocialLoginPassword": "somePassword",
    "loginUrl": "https://example/login",
    "homepage": "https://example"
  }
}
```

In this example, the `loginUrl` specifies to the plugin which URL it should navigate to, and by providing the task a config object which includes the loginSelector key to tell it which DOM element to click on it, it will be able to click and then use the google provided social creds to do the login.

**Comment by mirzamusic** [comment]
I made a PR which will resolve this, tested it locally and works fine
https://github.com/lirantal/cypress-social-logins/pull/22

**Comment by lirantal** [comment]
@michaelwschultz  could you check out the new version with @mirzamusic's popup support and see if this works for you?

**Comment by iparamey** [comment]
Can anybody give a link to a real tests with sign in as google user?

**Comment by lirantal** [comment]
Yes, check the CI here: https://github.com/lirantal/cypress-social-logins/pull/20
It fails because of a hard time to configure that on cloud but definitely should work locally if you replicate the same setup of the ./integrations/ folder

**Comment by SonaliQA** [comment]
On cypress new pop up I cannot enter email on Gmail login screen

**Comment by lirantal** [comment]
Can you share a video of what's happening?

**Comment by SonaliAvaya** [comment]
ya sure
please don't share to anyone other as this is confidential

*Video link
: https://drive.google.com/file/d/16RFh7W0WKT9omrrCTLlR1dwh5NV6U52n/view
<https://drive.google.com/file/d/16RFh7W0WKT9omrrCTLlR1dwh5NV6U52n/view>*

*Help me out what the other solutions? *
As my que is :
<https://stackoverflow.com/posts/61600656/timeline>

I have to perform below steps in cypress test I have user Page-Object
model, commands and env variable as well

1. Login with my app username and password

2. send an invitation to Gmail ID

3. Login into the Gmail with Gmail UN and password

4. Click on inbox and click on the first email which got the invitation
from app username

5. click on the confirmation email and then its navigate to the second tab
6. login with my app

Please help me with how I can test with these scenarios in the cypress test?

**Comment by Srinivasa1994** [comment]
Can you help me in being precise on what should we pass on login selector
![image](https://user-images.githubusercontent.com/56039595/81174610-c4b5dd80-8fbf-11ea-9db0-ab46c9a184f6.png)

![image](https://user-images.githubusercontent.com/56039595/81174709-f038c800-8fbf-11ea-970e-0a6fee3008a0.png)

**Comment by lirantal** [comment]
@SonaliAvaya 
1. looks like you're missing on providing the loginSelector
2. your overall process/config doesn't seem right. you shouldn't be browsing directly to google login but rather to your own site that requires it, then put the login selector url in that loginSelector  field

**Comment by SonaliQA** [comment]
[image: image.png]
[image: image.png]

index.js under plugins folder
[image: image.png]

cypress.json
[image: image.png]

Video link of executed test.js :
https://drive.google.com/file/d/14HxjrGdgq84CN91x1COia-qnJBDSxiVv/view
Its executing 3 times I don't know why? and could not select  first email
in inbox

I actually don't know what I should add now but I need multiple browser
support in cypress
I am new to cypress and need to login in cypress and go to inbox click on
particular email then confirm that invited link in email and navigate to
the different tab and then perform some actions on my app
I have tried a lot many times but I was not able to do so
please help me out : how I can add config and What I  should add in  login
selector

Thanks
Sonali

**Comment by SonaliQA** [comment]
and i dont know what i should pass on "loginselector" bcz I didctly want to
visit "gmail login " from goggle

**Comment by iparamey** [comment]
Please, can you check my issue from this video - https://cl.ly/f078a42ea763
Maybe you can give some advice

**Comment by pylnatahp** [comment]
I have the similar issue > TimeoutError: waiting for selector "input[type="email"]" failed: timeout 30000ms exceeded 
Popup opens but then nothing happens.

**Comment by lirantal** [comment]
I do hope you realize that your video, and credentials are publicly share here with everyone able to see it? I've already redacted your previous comments to avoid leaking of information so please don't repeat this unless you realize what you are sharing.

I saw that the video is in another locale than English so maybe the fields are different. Can you hover over the email input field with DevTools and tell me what is the class id for that?
![image](https://user-images.githubusercontent.com/316371/82180197-c0a38d00-98e8-11ea-9c20-3b088af564ee.png)

**Comment by iparamey** [comment]
It's not a problem because in my video you can see only test's credentials for the local environment. 
As for email input, I will attach a screenshot 
![Screenshot 2020-05-18 at 15 46 12](https://user-images.githubusercontent.com/17482045/82214638-d3d14f80-991e-11ea-9670-9b8ef7a4bcb5.png)

**Comment by lirantal** [comment]
That selector seems fine then with the code: https://github.com/lirantal/cypress-social-logins/blob/master/src/Plugins.js#L108

Could you add a popup delay? I think that's the issue.
You should add it in the socialLoginOptions variable you have. Give it a bit of time until it loads and try it.

**Comment by bmsoko** [comment]
Hello all! can anyone please share a hack/workaround to test other social networks? I need to click on other social networks, authenticate and authorize ! is there any way to do this via API or anything?

**Comment by iparamey** [comment]
> You should add it in the socialLoginOptions variable you have. Give it a bit of time until it loads and try it.

I'm sorry. But I don't understand how to add a popup delay :(

**Comment by iparamey** [comment]
> That selector seems fine then with the code: https://github.com/lirantal/cypress-social-logins/blob/master/src/Plugins.js#L108
> 
> Could you add a popup delay? I think that's the issue.
> You should add it in the socialLoginOptions variable you have. Give it a bit of time until it loads and try it.

Thanks a lot. It works now.

**Comment by haisumabbas** [comment]
Hi, 
I'm stuck at the same problem. It opens up the social login popup but doesn't type in the username.

![image](https://user-images.githubusercontent.com/18647257/83335918-739cbf00-a2c9-11ea-87f9-375245f68728.png)

![image](https://user-images.githubusercontent.com/18647257/83335941-a34bc700-a2c9-11ea-8dca-7d9ac7c5e855.png)

and for the other value of the headless flag, it returns an error for the other identifier/selector.

<input type="email" class="whsOnd zHQkBf" jsname="YPqjbf" autocomplete="username" spellcheck="false" tabindex="0" aria-label="Email or phone" name="identifier" value="" autocapitalize="none" id="identifierId" dir="ltr" data-initial-dir="ltr" data-initial-value="">

**Comment by lirantal** [comment]
Can you try adding a meaningful delay for the popup? it's one of the options you can pass to the plugin

**Comment by haisumabbas** [comment]
I tried different values for popupDelay but unfortunately, it's still the same.

**Comment by eledwinn** [comment]
This continues to happen.

I am sure the username and password are configured correctly.

my package.json file
```
{
  "name": "cypress-login",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "cypress": "cypress open"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "cypress": "^4.7.0",
    "cypress-social-logins": "^1.3.2"
  }
}
``` 

MacOS Mojave 10.14.6 (18G4032)
Node: v10.16.0
Npm: 6.14.4

**Comment by mumairofficial** [comment]
By adding `popupDelay` and `headless : false` configurations, I can see the popup and automated typing an email address. But there is another issue.

**Video link:**
https://www.loom.com/share/0b13b447f9c54a4e81beeee856e45544

**Screenshot:**
![image](https://user-images.githubusercontent.com/6498132/84723929-470ec580-afa0-11ea-91c6-576fc80e6850.png)

I have tried toggling `chromeWebSecurity` flag

Not sure if I am doing something wrong or google added extra browser security check,

Kindly help me out.

**Comment by TomaszWaszczyk** [comment]
```
    const username = Cypress.env('googleSocialLoginUsername')
    const password = Cypress.env('googleSocialLoginPassword')
    const loginUrl = Cypress.env('loginUrl')
    const cookieName = Cypress.env('cookieName')
```

Under those variables do you have real credentials to Gmail? `loginUrl` could be localhost? What should be value of `cookieName`?

**Comment by iledzka** [comment]
> By adding `popupDelay` and `headless : false` configurations, I can see the popup and automated typing an email address. But there is another issue.
> 
> **Video link:**
> https://www.loom.com/share/0b13b447f9c54a4e81beeee856e45544
> 
> **Screenshot:**
> ![image](https://user-images.githubusercontent.com/6498132/84723929-470ec580-afa0-11ea-91c6-576fc80e6850.png)
> 
> I have tried toggling `chromeWebSecurity` flag
> 
> Not sure if I am doing something wrong or google added extra browser security check,
> 
> Kindly help me out.

Log in to your gmail and turn on **Allow less secure apps** in Settings



> ```
>     const username = Cypress.env('googleSocialLoginUsername')
>     const password = Cypress.env('googleSocialLoginPassword')
>     const loginUrl = Cypress.env('loginUrl')
>     const cookieName = Cypress.env('cookieName')
> ```
> 
> Under those variables do you have real credentials to Gmail? `loginUrl` could be localhost? What should be value of `cookieName`?

loginUrl can be localhost, it's the url to your login page. I may be wrong, but I don't think that the cookieName matters that much. I managed to log it using arbitrary value in the cookieName.

**Comment by TomaszWaszczyk** [comment]
I solved it by implementation similar to: https://github.com/cypress-io/cypress-example-recipes/tree/master/examples/stubbing-spying__google-analytics

**Comment by mumairofficial** [comment]
> Log in to your gmail and turn on Allow less secure apps in Settings

Thank you @iledzka for the tip 👍In my opinion, there is a possibility that google might introduce new anti-bot configurations in future and that might break CICD pipeline for no reason.

Testing around 3rd party is anti pattern and I manage to bypass google auth with stubbing and spying techniques.

**Comment by marcovignotto** [comment]
Hi Everyone!
Looks like I'm not the only having problems!
Cypress does not find the button, I tried every kind of syntax (type, classes... and a cy.get("class") works) but it does not find the button to open the google login. I'm stuck here.

I always get: 
waiting for selector "input#identifierId[type="email"]" failed: timeout 30000ms exceeded

but it does not open the login so it will never find it.
Any idea?
Thanks!

**Comment by lindsaymacvean** [comment]
I was able to write a complete login process from AWS Cognito to Google Login pages and back to my app. Might be helpful to some...

[https://stackoverflow.com/a/74435696/2827300](https://stackoverflow.com/a/74435696/2827300)
